docs/internal/3i.md

```diff
@@ -0,0 +1,169 @@
+# 3i and Grates
+
+## ThreeI
+
+3i (Three-I) provides a general means for cages (and grates) to make system calls and intercept system calls via a programmable system call table.  The goal is to enable complex functionality without modifying the microvisor or increasing the microvisor's trusted computing base. 
+
+### Motivation
+
+In a traditional Linux environment, extending or intercepting system calls (e.g., adding a new filesystem, tracing, or filtering calls) requires kernel modifications or mechanisms such as ptrace, which incur heavy overhead and depend on kernel-level mediation.
+
+3i eliminates these constraints by introducing a user-space routing layer between cages, grates, and the underlying microvisor. Calls can be dispatched directly between grates or delegated to the microvisor when necessary, achieving kernel-level extensibility while keeping all new logic external to the kernel’s TCB.
+
+### High-level Concepts
+
+In traditional operating systems, a process makes a system call which traps into the kernel.  Every process which makes a system call ends up trapping into the same kernel routine.  In essence, there is one system call table which is shared by every process.
+
+In contrast, in Lind, 3i provides a per-cage system call table.  A cage can write a function that is a system call handler (or handler, for short) and then register a handler ( register_handler ) for a specific system call of a cage (security will be discussed later).  Whenever that cage makes a system call ( make_syscall ), it ends up calling that function.  Since every cage has its own system call table, the handlers for different cages may vary.
```

Comment:
> ...register a handler for a specific system call of a cage. Whenever that cage makes _a_ system call, it ends up calling that function.

(emphasis mine)

This is a bit unclear. Is there a handler per system call per cage or a single handler per cage?